### PR TITLE
Move ast collector template implementation to header file

### PR DESCRIPTION
**Pull request by SainiAditya1**
Thank you for making Rust GCC better!

If your PR fixes an issue, you can add "Fixes #issue_number" into this
PR description and the git commit message. This way the issue will be
automatically closed when your PR is merged. If your change addresses
an issue but does not fully fix it please mark it as "Addresses #issue_number"
in the git commit message.

Here is a checklist to help you with your PR.

- \[ ] GCC development requires copyright assignment or the Developer's Certificate of Origin sign-off, see https://gcc.gnu.org/contribute.html or https://gcc.gnu.org/dco.html
- \[ ] Read contributing guidlines
- \[ ] `make check-rust` passes locally
- \[ ] Run `clang-format`
- \[ ] Added any relevant test cases to `gcc/testsuite/rust/`

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.
---

*Please write a comment explaining your change. This is the message
that will be part of the merge commit.

### CODE REVIEW

**Review by CohenArthur — APPROVED** [review]

**Comment by CohenArthur** [comment]
Oh, @SainiAditya1: you'll also need to remove the implementation from `rust-ast-collector.cc`

**Comment by SainiAditya1** [comment]
@CohenArthur sure , I will remove .

File: gcc/rust/ast/rust-ast-collector.h [hunk 1]
```diff
@@ -47,15 +47,36 @@ class TokenCollector : public ASTVisitor
    */
   template <typename T>
   void visit_items_joined_by_separator (T &collection,
-					TokenId separator = COMMA,
-					size_t start_offset = 0,
-					size_t end_offset = 0);
```
Comment (P-E-P) [inline gcc/rust/ast/rust-ast-collector.h:52]: You need to keep default arguments, those are used in the visitor in many situations, removing them means the code won't compile correctly.

File: gcc/rust/ast/rust-ast-collector.h [hunk 2]
```diff
@@ -64,17 +85,41 @@ class TokenCollector : public ASTVisitor
...
    * with increased indentation. Also includes special handling for empty
    * collection to print only the delimiters with no new line inside.
    */
   template <typename T>
-  void visit_items_as_block (T &collection, std::vector<TokenPtr> trailing = {},
-			     TokenId left_brace = LEFT_CURLY,
-			     TokenId right_brace = RIGHT_CURLY);
+  void visit_items_as_block (T &collection, std::vector<TokenPtr> trailing ,
+			     TokenId left_brace ,
+			     TokenId right_brace )
```
Comment (P-E-P) [inline gcc/rust/ast/rust-ast-collector.h:102]: Same thing here, we need to keep those default values.

File: gcc/rust/ast/rust-ast-collector.h [hunk 3]
```diff
@@ -106,7 +151,10 @@ class TokenCollector : public ASTVisitor
   /**
    * @see visit<std::unique_ptr<T>>
    */
-  template <typename T> void visit (T &node);
+  template <typename T> void visit (T &node)
+  {
+  node.accept_vis (*this);
+  }
```
Comment (P-E-P) [inline gcc/rust/ast/rust-ast-collector.h:157]: I may be wrong but this part does not look correctly formatted ?

File: gcc/rust/ast/rust-ast-collector.h [hunk 4]
```diff
@@ -47,15 +47,36 @@ class TokenCollector : public ASTVisitor
...
+              }
+
+          }
 
   /**
    * Visit item placing end of line after.
    */
   template <typename T>
-  void visit_as_line (T &item, std::vector<TokenPtr> trailing = {});
+  void visit_as_line (T &item, std::vector<TokenPtr> trailing )
```
Comment (P-E-P) [inline gcc/rust/ast/rust-ast-collector.h:72]: We need to keep the empty vector default value.

**Review by P-E-P — CHANGES_REQUESTED** [review]
There are some modifications required for the code to successfully pass the ci:
- the code should compile, this could be easily fixed by adding back the default argument values
- the code should be correctly formatted
- every commit should compile independently. I suggest you squash all your commits together. It doesn't really matter if your commits are not atomic since gcc has a long tradition of monstrously large commits.

Keep the good work, it may seems tedious but you're not that far!

**Comment by philberty** [comment]
@SainiAditya1 you need to rebase your PR against master then fix the commit format and code formatting before we can merge this

**Comment by SainiAditya1** [comment]
@philberty

**Comment by philberty** [comment]
@SainiAditya1 your git commit messages are failing:

Checking 317c20c22bd0c943d59366c6c1ac8b5ff23a8882: OK
Checking b2e5dd04ea85d1db626d71424648e920263a2298: FAILED
ERR: cannot find a ChangeLog location in message
Checking c89a5b2eea37549eb4532b5eaba3c1842969faca: OK
Checking 7ddce9078a36e0bbcc08f5c28b400dff0e28d7d0: FAILED
ERR: cannot find a ChangeLog location in message

File: Makefile.def [hunk 5]
```diff
@@ -357,7 +357,6 @@ dependencies = { module=configure-gcc; on=all-ld; };
 dependencies = { module=configure-gcc; on=all-gold; };
 dependencies = { module=configure-gcc; on=all-libiconv; };
 dependencies = { module=all-gcc; on=all-libiberty; hard=true; };
-dependencies = { module=all-gcc; on=all-libgrust; };
```
Comment (philberty) [inline Makefile.def:360]: I dont think this should have changed this was a fix for parallel builds

File: Makefile.in [hunk 6]
```diff
@@ -67178,16 +67178,6 @@ all-stagetrain-gcc: all-stagetrain-libiberty
 all-stagefeedback-gcc: all-stagefeedback-libiberty
 all-stageautoprofile-gcc: all-stageautoprofile-libiberty
 all-stageautofeedback-gcc: all-stageautofeedback-libiberty
-all-gcc: maybe-all-libgrust
```
Comment (philberty) [inline Makefile.in:67181]: this is unrelated change

File: configure [hunk 7]
```diff
@@ -3439,17 +3439,6 @@ $as_echo "yes" >&6; }
     fi
 fi
 
-# Disable libgrust when not using the rust language
```
Comment (philberty) [inline configure:3442]: same

File: configure.ac [hunk 8]
```diff
@@ -698,17 +698,6 @@ if test -d ${srcdir}/libphobos; then
     fi
 fi
 
-# Disable libgrust when not using the rust language
```
Comment (philberty) [inline configure.ac:701]: same

File: gcc/testsuite/rust/compile/issue-2019-2.rs [hunk 9]
```diff
@@ -1,30 +0,0 @@
-#[lang = "add"]
-pub trait Add<RHS = Self> {
```
Comment (philberty) [inline gcc/testsuite/rust/compile/issue-2019-2.rs:2]: Why are you deleting this?

**Comment by philberty** [comment]
Hi @SainiAditya1, I am going to close this PR as it's very far off course from the initial issue and the deletion of this many test cases and build system changes are not correct.

I suggest starting from a fresh branch and making a new PR.

**Pull request closed without merging**